### PR TITLE
Add data stream support to the reindex api.

### PR DESCRIPTION
**Pull request by martijnvg**
Relates to #53100 and #57788

With this change reindex get index into a data stream if `op_type=create` is specified in the reindex request body. This is required otherwise a validation error occurs that data stream only accepts index request with create op type.

### CODE REVIEW

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/Reindexer.java [hunk 1]
```diff
@@ -265,7 +266,11 @@ protected void finishHim(Exception failure, List<BulkItemResponse.Failure> index
              */
             index.routing(mainRequest.getDestination().routing());
             index.setPipeline(mainRequest.getDestination().getPipeline());
-            // OpType is synthesized from version so it is handled when we copy version above.
+            // Only copy the op_type if it has been set to CREATE.
+            // (INDEX is the default and UPDATE and DELETE are invalid values)
+            if (mainRequest.getDestination().opType() == DocWriteRequest.OpType.CREATE) {
```
Comment (martijnvg) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/Reindexer.java:271]: According to the reindex docs, op_type is a valid argument. However it was currently not supported, as the op_type wasn't copied over from the destination index request template to the actual index requests that reindex api generates.

File: docs/reference/indices/rollover-index.asciidoc [hunk 2]
```diff
@@ -254,7 +254,7 @@ POST /my-data-stream/_rollover <2>
 --------------------------------------------------
 // TEST[continued]
 // TEST[setup:huge_twitter]
-// TEST[s/# Add > 1000 documents to my-data-stream/POST _reindex?refresh\n{"source":{"index":"twitter"},"dest":{"index":".ds-my-data-stream-000001"}}/]
+// TEST[s/# Add > 1000 documents to my-data-stream/POST _reindex?refresh\n{"source":{"index":"twitter"},"dest":{"index":"my-data-stream","op_type":"create"}}/]
```
Comment (martijnvg) [inline docs/reference/indices/rollover-index.asciidoc:257]: This change is required for #57788, but in order to that all other changes in this pr are needed.

**Review by danhermann — APPROVED** [review]
LGTM. This looks like it will automatically pick up all backing indices for a data stream which contradicts our recommendation to perform a separate reindex request per index, but if we need to address that, it would be a separate PR.

**Comment by martijnvg** [comment]
> This looks like it will automatically pick up all backing indices for a data stream which contradicts our recommendation to perform a separate reindex request per index, but if we need to address that, it would be a separate PR.

I think that is okay and intended? If a data stream is used as source index then all backing indices are resolved (happens via search api) and if a data stream is used as destination index then only the write index is resolved (via bulk api).

**Review by henningandersen — APPROVED** [review]
LGTM.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/Reindexer.java (hunk 1 above)
Comment (martijnvg) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/Reindexer.java:274]: That is true. I will remove it and include the it in the commit message.

**Pull request merged**